src/NHibernate/Driver/TableBasedBulkProvider.cs

```diff
@@ -0,0 +1,135 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Data;
+using System.Linq;
+using System.Web.UI;
+using NHibernate.Engine;
+using NHibernate.Id;
+using NHibernate.Persister.Entity;
+using NHibernate.Type;
+
+namespace NHibernate.Driver
+{
+	public abstract class TableBasedBulkProvider : BulkProvider
```

Comment:
We have a custom solution for Bulk Copy to SQL Server in our application framework generating an IDataReader out of an IEnumerable of entities.

We only support one target table not splitting the entities because of different mapping types but it has a huge advantage that we can build a streaming pipeline for a large set of entities insteadof saving everything in memory into a DataTable.

Don't know if and how that idea could be implemented based on that branch because I did not look into that code base in detail but that would be the most efficient kind of bulk import to an SQL Server database.